src/firmware/types.rs

```diff
@@ -202,3 +208,522 @@ pub struct SnpPlatformStatus {
     /// Reported TCB version.
     pub reported_tcb_version: TcbVersion,
 }
+
+#[repr(C, packed)]
+pub struct SnpConfig {
+    reported_tcb: u64,
+    mask_chip_id: u32,
+    reserved: [u8; 52],
+}
+
+impl Default for SnpConfig {
+    fn default() -> Self {
+        Self {
+            reported_tcb: Default::default(),
+            mask_chip_id: Default::default(),
+            reserved: [0; 52],
+        }
+    }
+}
+
+#[repr(C)]
+pub struct SnpSetExtConfig<'a> {
+    /// Address of the SnpConfig or 0 when reported_tcb does not need
+    /// to be updated.
+    pub config_address: Option<&'a SnpConfig>,
+
+    /// Address of extended guest request certificate chain or 0 when
+    /// previous certificate should be removed on SNP_SET_EXT_CONFIG.
+    pub certs_address: Option<*mut c_void>,
+
+    /// Length of the certificates.
+    pub certs_len: u32,
+}
+
+#[repr(C)]
+pub struct SnpGetExtConfig<'a> {
+    /// Address of the SnpConfig or 0 when reported_tcb should not be
+    /// fetched.
+    pub config_address: Option<&'a SnpConfig>,
+
+    /// Address of extended guest request certificate chain or 0 when
+    /// certificate should not be fetched.
+    pub certs_address: Option<*mut c_void>,
+
+    /// Length of the buffer which will hold the fetched certificates.
+    pub certs_len: u32,
+}
+
+/// Locates the non-zero based index bit field of a specified number.
+///
+/// `$field`: expr - The number in question the bit should be taken from.
+/// `$bit`: expr - The one-based index bit being requested.
+/// `$dt`: ty - The data type of the field (for alignment reasons).
+///
+/// # Examples:
+///
+/// To find the 3rd bit field of Guest Field Select of a Derived Key Request:
+///
+/// ```
+/// let mut key_req: DerivedKeyReq = Default::default();
+/// key_req.guest_field_select = 4; /// 0b0..100 the third bit should be on.
+/// let res = bit_val\!(key_req.guest_field_select, 3, u64);
+/// assert_eq!(res, 1);
+/// ```
+///
+#[macro_export]
+macro_rules! bit_val {
+    ($field: expr, $bit: expr, $dt: ty) => {
+        {
+            ($field & (1 as $dt << ($bit as $dt - 1 as $dt))) >> ($bit - 1 as $dt)
+        }
+    }
+}
+
+#[cfg(feature = "openssl")]
+fn swap_endian(bytes: &[u8]) -> Vec<u8> {
+    let mut retval: Vec<u8> = bytes.into();
+    retval.reverse();
+    retval
+}
+
+/// Information provided by the guest owner for requesting an attestation
+/// report from the AMD Secure Processor.
+#[repr(C)]
+pub struct ReportReq {
+    /// Guest-provided data to be included into the attestation report
+    pub report_data: [u8; 64],
+
+    /// The VMPL to put into the attestation report. Must be greater than or
+    /// equal to the current VMPL and at most three.
+    pub vmpl: u32,
+
+    /// Reserved memory slot, must be zero.
+    reserved: [u8; 28],
+}
+
+impl Default for ReportReq {
+    fn default() -> Self {
+        Self {
+            report_data: [0; 64],
+            vmpl: Default::default(),
+            reserved: Default::default(),
```

Comment:
`[0; 28]`, unless `Default::default()` always will zero-out this array.

Reply:
It is my understanding that `std::Default` always zeroes out the fields of an array, and that implementations are provided for arrays with a length of 32 or less.

[Related Reading](https://doc.rust-lang.org/src/core/default.rs.html#208)

Reply:
Great, this should work fine then.